### DISCUSSION TITLE
How to protect static folder on the backend(!) with authentication and still use it with QImg as :src

### DISCUSSION THREAD

**Discussion by Dirk-** [quasarframework/quasar#10968] [General - Components / Directives / etc] [answered]
I have a Quasar v1 SPA with a node.js backend which is connected to a MongoDB database. The database contains document files of all sorts. Some of them are images or pdfs which I want to display directly in the SPA, others can just be downloaded via the app.
The backend gets these file out of the database and stores them inside a temp folder in `/public` for the SPA to get them for download or to show them in a QImg component like this

`<q-img :src="document.url">`

`document.url` contains the complete URL to the file on the backend.

I use Passport authentication with Axios to communicate with the backend, but I did not protect the `/public` route yet. I tried to do that and included the Axios bearer in the fetch options for download. This works well, but of course QImg can't access the folder any more.

How can I add the authentication info to the `:src` attribute of QImg? Is it possible to provide a function somehow?

**Comment by suleimanas** [comment]
:src="base64 encoded image url"

**Comment by hawkeye64** [comment, marked as answer]
You are already using Axios, so use that to download the image with authentication and then set the QImg

**Reply by Dirk-** [reply]
@hawkeye64, @suleimanas You mean, I should write a function to

- `fetch()` the image from the backend with authentication, 
- get the `blob` from the `response` and then 
- use `URL.createObjectURL(blob)`
 
to get the URL for QImg?

Would `:src` accept an async function as a value?

I get `[Vue warn]: Invalid prop: type check failed for prop "src". Expected String, got Promise`, and I do not know how to collect all these async calls into a sync function and really return a string.

Maybe I can create the URL in the `created()` part of my script and use it later in the template.

**Reply by suleimanas** [reply]
1 & 2 = Yes fetch the image from your backend as a blob
3 = createObjectURL is bad for memory if you have multiple images use FileReader better
then assign that base64ImgData to the img :src

(i'm sorry i'll write from my memory and you 'll get the idea):
```
<template>
<q-img :src="imgVal" />
</template>
<script>
export default {
  data(){
    return {
      imgVal: 'any placeholder',
    }
  },
  created(){
    // fetch the blob
    const imgBlob = fetch(...);
     const imgFile = new FileReader();
     imgFile.readAsDataURL(imgBlob); 
    imgFile.onload = () => {
        imgVal= imgFile.result;                
    }
  }
}
</script>
```

**Reply by metalsadman** [reply]
3. https://forweb.dev/en/blog/2020-05-05-object-url/
https://stackoverflow.com/questions/31742072/filereader-vs-window-url-createobjecturl

**Reply by suleimanas** [reply]
https://www.andygup.net/samples/filereader-createobjurl-test.html

**Reply by metalsadman** [reply]
try it with large images

**Reply by suleimanas** [reply]
Indeed
https://codepen.io/sasg87962/pen/VwzeWax

**Reply by suleimanas** [reply]
agreed i can see it's faster in Firefox sometimes, i'm going to use it more in the future

**Reply by Dirk-** [reply]
Thank you all.
